validator/client/src/main/java/tech/pegasys/artemis/validator/client/DutyScheduler.java

```diff
@@ -69,13 +86,19 @@ public void onSlot(final UnsignedLong slotNumber) {
           for (UnsignedLong currentEpoch = startEpoch;
               currentEpoch.compareTo(endEpoch) <= 0;
               currentEpoch = currentEpoch.plus(ONE)) {
-            scheduleDutiesForEpoch(currentEpoch).reportExceptions();
+            scheduleDutiesForEpoch(currentEpoch);
           }
           return startEpoch.compareTo(endEpoch) > 0 ? lastRequestedEpoch : endEpoch;
         });
   }
 
-  private SafeFuture<Void> scheduleDutiesForEpoch(final UnsignedLong epoch) {
+  private void scheduleDutiesForEpoch(final UnsignedLong epoch) {
+    final SafeFuture<Void> future = requestAndScheduleDutiesForEpoch(epoch);
+    pendingTasksByEpoch.put(epoch, future);
+    removeWhenAllTasksComplete(epoch, future);
```

Comment:
This line assumes that this validator won't have any duties for this epoch, so it will need to remove the future above, right?

Reply:
`removeWhenAllTasksComplete` will only remove the entry if the key and value match.  So this is handling the normal case where we are scheduling duties ahead of time and it completes scheduling before those slots occur.  In that case when `future` completes it will still be the value in the map and so will be removed.  When the slots occur there's no entry in the map so they execute immediately.

If however while the request for duties is in-flight one of the slots for that epoch occurs, that slot will replace this future with the one for it's own task and this call to `removeWhenAllTasksComplete` will wind up doing nothing. Responsibility for cleaning up moves to the task that was scheduled by onSlot (and this process repeats for each scheduled task - it gets tacked onto the end and takes responsibility for removing things).

Which I think is a long winded way of saying yes. :)

Reply:
Thanks for the detailed answer :)